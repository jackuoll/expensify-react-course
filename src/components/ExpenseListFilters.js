import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(()=> ({ calendarFocused }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortSelectChange = (e) => {
    if (e.target.value === "date"){
      this.props.sortByDate();
    } else if (e.target.value === "amount") {
      this.props.sortByAmount();
    }
  };
  render() {
    return (
      <div>
        <input type="text" defaultValue={this.props.filters.text} onChange={this.onTextChange} />
        <select defaultValue={this.props.filters.sortBy} onChange={this.onSortSelectChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={false}
          numberOfMonths={1}
          isOutsideRange={() => false}
          minimumNights={0}
        />
      </div>
    )
  }
};

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (value) => dispatch(setTextFilter(value)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setStartDate: (value) => dispatch(setStartDate(value)),
  setEndDate: (value) => dispatch(setEndDate(value))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
