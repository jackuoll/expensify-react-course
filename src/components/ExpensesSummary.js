import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expensesTotal';

export const ExpensesSummary = (props) => {
  const quantity = props.expenses.length;
  const expenseWording = quantity === 1 ? 'expense' : 'expenses';
  if (quantity === 0) {
    return null;
  }
  const totalCost = numeral(getExpensesTotal(props.expenses) / 100).format('$0,0.00');
  return (
    <div>
      <h1>Viewing {quantity} {expenseWording} totalling {totalCost}.</h1>
    </div>
  );
};

const mapStateToProps = (state) => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);
