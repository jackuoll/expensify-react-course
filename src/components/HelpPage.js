import React from 'react';

const HelpPage = () => (
  <div>
    This is some help.
  </div>
);

export default HelpPage;