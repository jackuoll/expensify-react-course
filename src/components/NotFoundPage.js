import React from 'react';
import { Link } from 'react-router-dom'

const NotFoundPage = () => (
  <div>
    404!
    <p><Link to="/">Go back to home.</Link></p>
  </div>
);

export default NotFoundPage;