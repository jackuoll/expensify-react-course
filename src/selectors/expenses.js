import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = isNaN(startDate) || createdAtMoment.isSameOrAfter(moment(startDate), 'day');
    const endDateMatch = isNaN(endDate) || createdAtMoment.isSameOrBefore(moment(endDate), 'day');
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy == 'date'){
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy == 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
    return 0;
  });
};
