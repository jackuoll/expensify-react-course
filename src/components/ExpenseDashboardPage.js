import React from 'react';
import ExpenseList from './ExpenseList';
import ExpensesSummary from './ExpensesSummary';
import ExpenseListFilters from './ExpenseListFilters';
import LoginPage from './LoginPage'

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpensesSummary />
    <ExpenseList /> 
  </div>
);

export default ExpenseDashboardPage;