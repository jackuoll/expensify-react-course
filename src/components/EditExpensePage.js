import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
  removeExpense = () => {
    this.props.removeExpense(this.props.expense);
    this.props.history.push('/');
  };
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Editing {this.props.expense.description}</h1>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit} 
        />
        <button onClick={this.removeExpense}>Remove</button>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  removeExpense: (expense) => dispatch(removeExpense(expense)),
  editExpense: (id, expense) => dispatch(editExpense(id, expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);