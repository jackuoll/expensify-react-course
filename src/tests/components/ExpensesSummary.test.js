import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('test that a single expense renders correctly', () => {
  const rendered = shallow(<ExpensesSummary expenses={[expenses[0]]} />);
  expect(rendered).toMatchSnapshot();
});

test('test that some expenses render correctly', () => {
  const rendered = shallow(<ExpensesSummary expenses={expenses} />);
  expect(rendered).toMatchSnapshot();
});
