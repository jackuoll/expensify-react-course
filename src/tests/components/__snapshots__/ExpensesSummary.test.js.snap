// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test that a single expense renders correctly 1`] = `
<div>
  Viewing 
  1
   
  expense
   totalling 
  $195.11
  .
</div>
`;

exports[`test that some expenses render correctly 1`] = `
<div>
  Viewing 
  3
   
  expenses
   totalling 
  $1,335.11
  .
</div>
`;
